libmodernize/main.py

```diff
@@ -17,6 +18,52 @@
 from libmodernize import __version__
 from libmodernize.fixes import lib2to3_fix_names, six_fix_names, opt_in_fix_names
 
+
+PY3K = sys.version_info >= (3, 0)
+if not PY3K:
+    LF = '\n'
+    CRLF = '\r\n'
+    CR = '\r'
+else:
+    LF = bytes('\n', encoding='ascii')
+    CRLF = bytes('\r\n', encoding='ascii')
+    CR = bytes('\r', encoding='ascii')
+
+
+class LFPreservingRefactoringTool(StdoutRefactoringTool):
+    """ https://github.com/python-modernize/python-modernize/issues/121 """
+    def write_file(self, new_text, filename, old_text, encoding):
+        # detect linefeeds
+        oldfile = open(filename, 'rb')
+        lineends = {LF:0, CRLF:0, CR:0}
+        for line in oldfile:
+            if line.endswith(CRLF):
+                lineends[CRLF] += 1
+            elif line.endswith(LF):
+                lineends[LF] += 1
+            elif line.endswith(CR):
+                lineends[CR] += 1
+        oldfile.close()
+        super(LFPreservingRefactoringTool, self).write_file(
+            new_text, filename, old_text, encoding)
+        # detect if line ends are consistent in source file
+        if sum([bool(lineends[x]) for x in lineends]) == 1:
+            # detect if line ends are different from system-specific
+            newline = [x for x in lineends if lineends[x] != 0][0]
+            if os.linesep != newline:
+                # rereading new file is easier that writing new_text
+                # correct encoding in Python 2 and 3 compatible way
```

Comment:
I've made this code without even running it, and because I can't wrap my head around how those encodings should behave in 2 and 3, I think this will be a complication. I don't like rereading all the files three times myself, but 1. python-modernize is not made for speed and 2. it is a job of disk cache